### PR TITLE
Serializing connections objects through NetworkAPI

### PR DESCRIPTION
**Pull request by Zbysekz**
Draft.

Hello @dkeeney,
Thanks to the feature that you develop with json parameters, pandaVis is now able to generate this automatically by using NetworkAPI
[https://www.youtube.com/watch?v=tXJtgsHOyO8](https://www.youtube.com/watch?v=tXJtgsHOyO8)
After i finish it to some stable version, i will publish how-to.

Now i am diving into maybe more complicated thing - the connections.
If you remember, for previous non-networkAPI version i did by your advice serialization of whole TM object to binary dump file (by using tm.saveToFile()) for each iteration, and afterwards i load distal connections specifically when requested from user.

I would need to do similar procedure with networkAPI.

So basically i need to have access to connections object of regions. (some of them like ApicalPair have two of them - basal and apical)

I made draft, it works for getting C++ region connections from python side.
e.g. `myConnections = net.getRegion('tm').getConnections("") # region type is TMRegion(C++)` 

Now i need to make it work also for getting registered python regions. e.g. to be able to call this for `py.ApicalTMPairRegion`

I got little stuck, i have some guess that i need to code in `/bindings/py//cpp_src/plugin/PyBindRegion.cpp` but i am not sure about that.

Maybe @breznak or someone else would know, i am not sure who is more "pro" with the bindings.

Thanks for any hint :)

### CODE REVIEW

**Comment by Zbysekz** [comment]
> It would write the serialization to a file as BINARY (which would not help you very much) or as JSON or XML which might be useful for you. 

The BINARY is fine for me, i serialize connections while baking the simulation, then in HTMpanda deserialize it and call get functions to get desired internal data with convinient functions like getSegments, getSynapseData etc..

> You show a print( ) in your example. That would be the streaming in C++ and that does exist for most C++ modules. It is intended to be read by humans and might not be very easy to parse.

These print() in this rough scratchbook code are here just for debugging, i just wanted to show code that i use for exploring what i can get from python instances. Just for case that you wanted to try it out also.

> Another idea to consider... the engine implements "ExecuteCommand". 

Yes this is it. Nice.

I tried to implement new command `SaveConnectionsToFile(filePath)` and it was successful. For c++ it was easy.
For python it seemed hard, but in the end it was far easier :D

Network API is build nicely: When you call `ExecuteCommand(methodName, args..)` it goes to c++ abstract region class, from where it calls implementation. If it is python implementation, it simply calls the python implementation method if exists.
So only thing i needed to do was to add function `SaveConnectionsToFile(filePath)` into each python implementation class.(ApicalTMRegion, ColumnPoolerRegion, GridCellRegion...)

See code change, i have now just part of it done.

File: bindings/py/cpp_src/bindings/engine/py_Region.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ namespace htm_ext
 		// This would be .py code calling getSpec() to get the Spec structure
 		// on an already instantiated class.
         py_Region.def("getSpec", &Region_t::getSpec);
-
+        py_Region.def("executeCommand", &Region_t::executeCommand);
```
Comment (Zbysekz) [inline bindings/py/cpp_src/bindings/engine/py_Region.cpp:53]: Note that this was the crucial thing. Exposing executeCommand to python.

File: src/test/unit/regions/TMRegionTest.cpp [hunk 2]
```diff
@@ -328,7 +328,7 @@ TEST(TMRegionTest, testLinking) {
   EXPECT_EQ(r4InputArray, expected3outa) << r4InputArray;
 
   // cleanup
-  region3->executeCommand({"closeFile"});
+  region4->executeCommand({"closeFile"});
```
Comment (Zbysekz) [inline src/test/unit/regions/TMRegionTest.cpp:331]: I accidentaly discovered this bug in test. Closing file was for TMRegion but should be for FileOutputRegion obviously. Before there was no code in TMRegion ExecuteCommand() so nothing was thrown.

File: bindings/py/cpp_src/bindings/engine/py_Region.cpp (hunk 1 above)
Comment (dkeeney) [reply bindings/py/cpp_src/bindings/engine/py_Region.cpp:53]: Oh, it looks like I missed that one when setting up the bindings.  Good catch.

File: src/test/unit/regions/TMRegionTest.cpp (hunk 2 above)
Comment (dkeeney) [reply src/test/unit/regions/TMRegionTest.cpp:331]: Another good catch. :)

Comment (dkeeney) [reply src/test/unit/regions/TMRegionTest.cpp:331]: Hmmm, maybe it should throw an exception if a command is requested but it is not implemented.
In RegionImpl.cpp line 210
```
std::string RegionImpl::executeCommand(const std::vector<std::string> &args, Int64 index) {
  // This Region does not execute any Commands.
  return "";
}
```
Should be something like ...
```
std::string RegionImpl::executeCommand(const std::vector<std::string> &args, Int64 index) {
  std::string command;
  if (args.size() > 0)
    command = args[0];

  // This Region does not execute any Commands.
  NTA_THROW << getName() << ": Does not implement the command  '" << command << "'";
}```

Can you include this in your PR  (along with a simple unit test)?

Comment (Zbysekz) [reply src/test/unit/regions/TMRegionTest.cpp:331]: Ok, sure. It is there

**Review by breznak — COMMENTED** [review]
Sorry, I just got back and haven't read the whole proposal fully, just want to remind the internal status of Connections: 

> Now i am diving into maybe more complicated thing - the connections.
If you remember, for previous non-networkAPI version i did by your advice serialization of whole TM object to binary dump file (by using tm.saveToFile()) for each iteration, and afterwards i load distal connections specifically when requested from user.
I would need to do similar procedure with networkAPI.
So basically i need to have access to connections object of regions. (some of them like ApicalPair have two of them - basal and apical) 

Our "standardized APIs" are for NetworkAPI and cpp/py "main classes"-TM,SP,.. which we try not to break for users. 
Connections is intended as internal structure which may change anytime (not that it does drastically, to be honest). 

But don't build any API on Connections and expect it not to break. 

Serializing and accessing connections is imho a possible way, if you can, don't expose Connections through the NetworkAPI's public methods, or mark it with !!!This can break!!!.

**Comment by Zbysekz** [comment]
> Our "standardized APIs" are for NetworkAPI and cpp/py "main classes"-TM,SP,.. which we try not to break for users.
> Connections is intended as internal structure which may change anytime (not that it does drastically, to be honest).
> 
> But don't build any API on Connections and expect it not to break.
> 
> Serializing and accessing connections is imho a possible way, if you can, don't expose Connections through the NetworkAPI's public methods, or mark it with !!!This can break!!!.

I just added "command" to all regions that have connections -> "saveConnectionsToFile(path)" so i think even we change the connection iternally, serialization will be there always.
HTM panda uses connections functions, so if we change connections in htm.core i would have to change it also in panda. But that is not problem for me. I also want to anchor some stable release of htm.core to panda requirements, that i don't have to worry about this much, and do updates myself regulary.

**Comment by Zbysekz** [comment]
To sum up changes:

- I added executeCommand for python regions
- created region "command" SaveConnectionsToFile(path) for all regions that have connections (i think except GridCellLocationLayer) - this serializes connections to dump file in binary form
- Also created hotgym example for network API - i think that this is first python networkAPI example in the code what i am aware of (FYI @dkeeney)


For hotgym example(or whatever else script using NetworkAPI) 
PandaVis automatically generates everything just by changing one import line
`from htm.bindings.engine_internal import Network`
to
`from pandaBaker.pandaNetwork import Network`

**Comment by breznak** [comment]
>> [don't depend directly on Connections stable API 

> I just added "command" to all regions that have connections -> "saveConnectionsToFile(path)" so i think even we change the connection iternally, serialization will be there always. 

ok, that's fine solution then, no problems with that :+1: 

Versioned htm.core dependency for pandaVis is good idea too.

File: py/htm/advanced/regions/ApicalTMPairRegion.py [hunk 3]
```diff
@@ -522,3 +522,12 @@ def getAlgorithmInstance(self):
         Returns an instance of the underlying temporal memory instance
         """
         return self._tm
+
+    def saveConnectionsToFile(self, filepath):
+        """
+        Binary dumps connections objects into file specified, adding suffix specifying the type
+        """
+        with open(filepath+"_basal.dump", "wb") as f:
+            f.write(self._tm.basalConnections.save())
```
Comment (breznak) [inline py/htm/advanced/regions/ApicalTMPairRegion.py:531]: nit: you could use the `connections.saveToFile("myfile.dump")`

File: py/htm/advanced/regions/ApicalTMPairRegion.py [hunk 4]
```diff
@@ -522,3 +522,12 @@ def getAlgorithmInstance(self):
         Returns an instance of the underlying temporal memory instance
         """
         return self._tm
+
+    def saveConnectionsToFile(self, filepath):
```
Comment (breznak) [inline py/htm/advanced/regions/ApicalTMPairRegion.py:526]: why not using the `executeCommand` here as well?

**Review by breznak — COMMENTED** [review]
I really like this, good job :+1: 
The NAPI Hello example is great! 
A few suggestions below

File: py/htm/advanced/regions/ApicalTMPairRegion.py (hunk 4 above)
Comment (Zbysekz) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:526]: This method is called by executeCommand.

Python implemented regions works like following: if you call executeCommand("whatever") on pyRegion, it natively tries to call method "whatever" of pythonRegion implementation (firstly checks if that attribute exists).

I think this is nice property of these python regions.

Comment (breznak) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:526]: > executeCommand("whatever") on pyRegion, it natively tries to call method "whatever" of pythonRegion 

oh, I see. I've seen you implementing the "saveConnectionsToFile" in the c++ SPRegion. Didn't know it works differently for PyRegions. thank you for explanation. 
Btw, through some reflection we could implement the same approach for cpp regions, I think? But may not be worth the problems as the if-else in executeCommand in cpp works well.

File: py/htm/advanced/regions/ApicalTMPairRegion.py (hunk 3 above)
Comment (Zbysekz) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:531]: I can't, this is not exposed to python..

i tried just now adding to bindings py_Connections.cpp
`py_Connections.def("saveToFile", &Connections::saveToFile);`

But i got 
```
TypeError: saveToFile(): incompatible function arguments. The following argument types are supported:
    1. (self: htm.bindings.algorithms.Connections, arg0: str, arg1: htm.bindings.engine_internal.SerializableFormat) -> None

Invoked with: <htm.advanced.algorithms.connections.Connections object at 0x7efed331d5f0>, '/home/ub/HTM/htm-2d-object-modeling/python/bakedDatabase/pandaVis_dumpData/L2_0_0_proximal.dump'
```

Because enum of the serializable format is not exposed to python, and that i dunno how to do

**Comment by breznak** [comment]
I'll wait for David, but this looks good to me :+1:

File: py/htm/advanced/regions/ApicalTMPairRegion.py (hunk 4 above)
Comment (Zbysekz) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:526]: > Btw, through some reflection we could implement the same approach for cpp regions, I think? But may not be worth the problems as the if-else in executeCommand in cpp works well.

Maybe possible, but as you say, for c++ i would keep as-is

File: py/htm/advanced/regions/ApicalTMPairRegion.py (hunk 3 above)
Comment (breznak) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:531]: it's possible to add enum to pybind, I've done it for ANOMALY_MODE , should be either in TM or Anomaly bindings. But then again, the current works quite well, so maybe not worth.

Comment (dkeeney) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:531]: An app MUST NOT call functions on a region implementations directly but you can use the ExecuteCommand facility in Region to call connections.saveToFile( ).  Calling directly will break the plugin mechanism which needs to work for both Python and C++ on all platforms.  If you are calling saveConnectionsToFile( ) in the python region using the "ExecuteCommand" than this is ok.  I just cannot tell how you are calling it.

By the way, the enums for save( ) and saveToFile( ) are exposed. See htmE\bindings/py/cpp_src/bindings/engine/py_Engine.cpp line 385.   Also, saveToFile( ) without the enum should default to BINARY.   Does that not work?  I also see that the save( ) with the enum was not implemented in the bindings...don't know why.

File: py/htm/examples/networkAPI/hotgym.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
+# ------------------------------------------------------------------------------
```
Comment (dkeeney) [inline py/htm/examples/networkAPI/hotgym.py:1]: Very nice example.

I would have used the Network.configure( ) function instead of using the 'default_parameters' and the addRegion( ) and Link( ) calls. It might be simpler but this works too.

**Review by dkeeney — APPROVED** [review]

File: py/htm/advanced/regions/ApicalTMPairRegion.py (hunk 3 above)
Comment (Zbysekz) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:531]: @dkeeney 
> An app MUST NOT call functions on a region implementations directly but you can use the ExecuteCommand facility in Region to call connections.saveToFile( ). Calling directly will break the plugin mechanism which needs to work for both Python and C++ on all platforms. 

I am using executeCommand in Panda, never calling directly, because i treat py regions in same way as cpp regions.

> If you are calling saveConnectionsToFile( ) in the python region using the "ExecuteCommand" than this is ok. I just cannot tell how you are calling it.

I believe that the code responsible for this nice behaviour is here: 
[https://github.com/htm-community/htm.core/blob/56753b0d9ab7e9e17ba837089ee1c8343d80f72f/bindings/py/cpp_src/plugin/PyBindRegion.cpp#L631](https://github.com/htm-community/htm.core/blob/56753b0d9ab7e9e17ba837089ee1c8343d80f72f/bindings/py/cpp_src/plugin/PyBindRegion.cpp#L631)
It basically calls whatever python method and parse result as string.

> By the way, the enums for save( ) and saveToFile( ) are exposed. See htmE\bindings/py/cpp_src/bindings/engine/py_Engine.cpp line 385. Also, saveToFile( ) without the enum should default to BINARY. Does that not work? I also see that the save( ) with the enum was not implemented in the bindings...don't know why.

It reports the error i mentioned above, i must supply serialization type... 
I also don't know. But i not solve this for now, since it is low priority and works like this with save() method.

Comment (dkeeney) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:531]: I agree, fixing the enums is low priority. But would you make an issue entry for that so we don't forget about it.  Thanks.

Comment (Zbysekz) [reply py/htm/advanced/regions/ApicalTMPairRegion.py:531]: Ok created https://github.com/htm-community/htm.core/issues/888
